### PR TITLE
Project purge command

### PR DESCRIPTION
**Pull request by brummett**
This adds a command for purging/expunging software results given one or more analysis projects.

It's based on jwalker's basic script, converted to a genome model command, and added locking around each anp to prevent multiple instances from stomping on each other.

### CODE REVIEW

File: lib/perl/Genome/Model/Command/Admin/PurgeSoftwareResultsFromAnalysisProject.pm [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+
+        $software_result_count++;
+        $total_kb_purged += $data->{kilobytes_requested};
+
+        if ($self->dry_run) {
+            $self->warning_message('Dry run, not removing software result '.$sr->id);
+        } else {
+            $self->status_message($reason);
+            $sr->expunge($reason);
+            UR::Context->commit();
```
Comment (davidlmorton) [inline lib/perl/Genome/Model/Command/Admin/PurgeSoftwareResultsFromAnalysisProject.pm]: Do we want some error-checking around this `Context->commit`?

File: lib/perl/Genome/Model/Command/Admin/PurgeSoftwareResultsFromAnalysisProject.pm [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+        }
+    }
+}
```
Comment (ebelter) [inline lib/perl/Genome/Model/Command/Admin/PurgeSoftwareResultsFromAnalysisProject.pm]: Need a '1;'

File: lib/perl/Genome/Model/Command/Admin/PurgeSoftwareResultsFromAnalysisProject.pm [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+    my $unlock_and_print_report = Scope::Guard->new(sub {
+        $unlocker->();
+        $self->status_message('Removed %d GB from %d software results',
+                                int($total_kb_purged / KB_IN_ONE_GB),
+                                $software_result_count);
+    });
+
+    my $sth = $dbh->prepare($sql);
+    die $self->error_message("preparing SQL failed : $DBI::errstr") unless $sth;
+
```
Comment (ebelter) [inline lib/perl/Genome/Model/Command/Admin/PurgeSoftwareResultsFromAnalysisProject.pm]: Maybe prepare the SQL statement handle outside the purging AnP method? And then reuse the $sth to execute?

File: lib/perl/Genome/Model/Command/Admin/PurgeSoftwareResultsFromAnalysisProject.pm (hunk 1 above)
Comment (ebelter) [inline lib/perl/Genome/Model/Command/Admin/PurgeSoftwareResultsFromAnalysisProject.pm]: Maybe each AnP purge should be in a transaction?

File: lib/perl/Genome/Model/Command/Admin/PurgeSoftwareResultsFromAnalysisProject.pm [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+        my $sr = Genome::SoftwareResult->get($data->{result_id});
+        die $self->error_message('Cannot get software result '. $data->{result_id}) unless $sr;
+
+        my $reason = 'Expunge software result uniquely used by model from disabled config item ('. $data->{profile_item_id} .') for analysis project \''. $data->{anp_name} .'\' ('. $data->{anp_id} .')';
+
+        $software_result_count++;
+        $total_kb_purged += $data->{kilobytes_requested};
+
+        if ($self->dry_run) {
+            $self->warning_message('Dry run, not removing software result '.$sr->id);
```
Comment (ebelter) [inline lib/perl/Genome/Model/Command/Admin/PurgeSoftwareResultsFromAnalysisProject.pm]: What about having the same message with prepended with 'Would remove' and Removing', like 'git clean'? That would you can see what would happen on a 'wet_run' :smile_cat:

Comment (ebelter) [inline lib/perl/Genome/Model/Command/Admin/PurgeSoftwareResultsFromAnalysisProject.pm]: I suppose it isn't as simple to just set UR_DBI_NO_COMMIT=1 for the dry_run.

Comment (brummett) [inline lib/perl/Genome/Model/Command/Admin/PurgeSoftwareResultsFromAnalysisProject.pm]: No, ```expunge()``` (and the things it calls) don't respect UR_DBI_NO_COMMIT before removing files.

**Comment by brummett** [comment]
Added new commits to die if the commit() fails, and unifying the messages printed while running

**Comment by brummett** [comment]
Added new commits to extract the SQL prepare, and to only purge data if --dry-run is off

**Comment by ebelter** [comment]
+1

**Comment by jeldred** [comment]
+1 :shipit:

**Pull request merged**